core/src/main/scala/kafka/server/KafkaRequestHandler.scala

```diff
@@ -146,64 +146,65 @@ class BrokerTopicMetrics(name: Option[String]) extends KafkaMetricsGroup {
-  def totalProduceRequestRate = metricTypeMap.getAndMaybePut(BrokerTopicStats.TotalProduceRequestsPerSec,
-    newMeter(BrokerTopicStats.TotalProduceRequestsPerSec, "requests", TimeUnit.SECONDS, tags))
-  def totalFetchRequestRate = metricTypeMap.getAndMaybePut(BrokerTopicStats.TotalFetchRequestsPerSec,
-    newMeter(BrokerTopicStats.TotalFetchRequestsPerSec, "requests", TimeUnit.SECONDS, tags))
-  def fetchMessageConversionsRate = metricTypeMap.getAndMaybePut(BrokerTopicStats.FetchMessageConversionsPerSec,
-    newMeter(BrokerTopicStats.FetchMessageConversionsPerSec, "requests", TimeUnit.SECONDS, tags))
-  def produceMessageConversionsRate = metricTypeMap.getAndMaybePut(BrokerTopicStats.ProduceMessageConversionsPerSec,
-    newMeter(BrokerTopicStats.ProduceMessageConversionsPerSec, "requests", TimeUnit.SECONDS, tags))
+  def failedProduceRequestRate = metricTypeMap(BrokerTopicStats.FailedProduceRequestsPerSec).meter()
+  def failedFetchRequestRate = metricTypeMap(BrokerTopicStats.FailedFetchRequestsPerSec).meter()
+  def totalProduceRequestRate = metricTypeMap(BrokerTopicStats.TotalProduceRequestsPerSec).meter()
+  def totalFetchRequestRate = metricTypeMap(BrokerTopicStats.TotalFetchRequestsPerSec).meter()
+  def fetchMessageConversionsRate = metricTypeMap(BrokerTopicStats.FetchMessageConversionsPerSec).meter()
+  def produceMessageConversionsRate = metricTypeMap(BrokerTopicStats.ProduceMessageConversionsPerSec).meter()
 
   // this method helps check with metricTypeMap first before deleting a metric
-  def removeMetricHelper(metricType: String, tags: scala.collection.Map[String, String]): Unit = {
-    val metric: Meter = metricTypeMap.remove(metricType)
-    if (metric != null) {
-      removeMetric(metricType, tags)
-    }
-  }
+  def removeMetricHelper(metricType: String, tags: scala.collection.Map[String, String]): Unit =
+    metricTypeMap.remove(metricType).foreach(_.close())
 
-  def close() {
-    removeMetricHelper(BrokerTopicStats.MessagesInPerSec, tags)
-    removeMetricHelper(BrokerTopicStats.BytesInPerSec, tags)
-    removeMetricHelper(BrokerTopicStats.BytesOutPerSec, tags)
-    removeMetricHelper(BrokerTopicStats.BytesRejectedPerSec, tags)
-    if (replicationBytesInRate.isDefined)
-      removeMetricHelper(BrokerTopicStats.ReplicationBytesInPerSec, tags)
-    if (replicationBytesOutRate.isDefined)
-      removeMetricHelper(BrokerTopicStats.ReplicationBytesOutPerSec, tags)
-    removeMetricHelper(BrokerTopicStats.FailedProduceRequestsPerSec, tags)
-    removeMetricHelper(BrokerTopicStats.FailedFetchRequestsPerSec, tags)
-    removeMetricHelper(BrokerTopicStats.TotalProduceRequestsPerSec, tags)
-    removeMetricHelper(BrokerTopicStats.TotalFetchRequestsPerSec, tags)
-    removeMetricHelper(BrokerTopicStats.FetchMessageConversionsPerSec, tags)
-    removeMetricHelper(BrokerTopicStats.ProduceMessageConversionsPerSec, tags)
-  }
+  def close() = metricTypeMap.values.foreach(_.close())
```

Comment:
I wasn't sure if we should clear the map here or not.
If we do clear it, there's no going back - nothing will repopulate it.
If we don't, we are saving metric in memory for no reason.

I traced the usages of #close() and they seem to be called when we either deallocate the `BrokerTopicMetrics` instance or shut down the server, so I think it's fine to not clear it. Nevertheless, I'm open to suggestions

Reply:
I don't think we should clear it for the reason I mentioned in the other comment. Or you'd have to add some logic to re-add to the underlying map. But the memory usage of the wrapper without the underlying metric is pretty small.

Reply:
With leader partition migrating back and forth, would that cumulate a lot of metrics no-longer-needed?

Reply:
Metrics are closed. We just keep the wrapper.

Reply:
I misunderstood the original question and thought @stanislavkozlovski was asking about the `foreach(_.close())` call.

Agree that we do not need to clear the map of wrappers.